From 0aaf4014589137553e09de3da5f9b20556c8ae01 Mon Sep 17 00:00:00 2001
From: Meticulus <theonejohnnyd@gmail.com>
Date: Fri, 23 Mar 2018 07:23:05 -0500
Subject: [PATCH] "0001-MediaCodec-use-google-h264-encoder
 _______________________________

*Previous From 1552b99fb04085fb259413af4efa900dbf64b090 Mon Sep 17 00:00:00 2001
*Previous From: Meticulus <theonejohnnyd@gmail.com>
*Previous Date: Sun, 7 May 2017 06:14:35 -0500
*Previous Subject: [PATCH] 0001-MediaCodec-use-google-h264-encoder
*
*Change-Id: I55b4c62f1ee5db7299284306613ea5d168b2adc2"

Change-Id: Id440d78decb725978c7d51acea0c44c2051a35cd
---
 media/libstagefright/MediaCodec.cpp | 29 +++++++++++++++++++++++++++--
 1 file changed, 27 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/MediaCodec.cpp b/media/libstagefright/MediaCodec.cpp
index 759e42d..ae35650 100644
--- a/media/libstagefright/MediaCodec.cpp
+++ b/media/libstagefright/MediaCodec.cpp
@@ -58,6 +58,9 @@
 #include <utils/Log.h>
 #include <utils/Singleton.h>
 
+#include <cutils/properties.h>
+#include <string.h>
+
 namespace android {
 
 // key for media statistics
@@ -416,7 +419,22 @@ sp<MediaCodec> MediaCodec::CreateByType(
         uid_t uid) {
     sp<MediaCodec> codec = new MediaCodec(looper, pid, uid);
 
-    const status_t ret = codec->init(mime, true /* nameIsType */, encoder);
+    /* Meticulus:
+     * The hw video encoder does not do well with screenrecording
+     * and the google decoder does not do well with 1080p video
+     * playback. So here we are forcing the use of the google
+     * encoder for video/avc for recording While retaining the
+     * default behavior for video playback.
+     * UPDATE: Lets make this switchable.
+     */
+    status_t ret = -1;
+    //ALOGD("Meticulus: mime = %s",mime.c_str());
+    if(!strcmp(mime.c_str(),"video/avc") && encoder && property_get_bool("persist.sys.google_avc_enc",0)) {
+	ret = codec->init("OMX.google.h264.encoder",false,encoder);
+	ALOGI("persist.sys.google_avc_enc is set.");
+    }
+    else
+	ret = codec->init(mime, true /* nameIsType */, encoder);
     if (err != NULL) {
         *err = ret;
     }
@@ -428,7 +446,14 @@ sp<MediaCodec> MediaCodec::CreateByComponentName(
         const sp<ALooper> &looper, const AString &name, status_t *err, pid_t pid, uid_t uid) {
     sp<MediaCodec> codec = new MediaCodec(looper, pid, uid);
 
-    const status_t ret = codec->init(name, false /* nameIsType */, false /* encoder */);
+    status_t ret = -1;
+    //ALOGD("Meticulus: name = %s",name.c_str());
+    if(!strcmp(name.c_str(),"OMX.IMG.TOPAZ.VIDEO.Encoder") && property_get_bool("persist.sys.google_avc_enc",0)) {
+	ret = codec->init("OMX.google.h264.encoder",false,true);
+	ALOGI("persist.sys.google_avc_enc is set.");
+    }
+    else
+	ret = codec->init(name, false /* nameIsType */, false /* encoder */);
     if (err != NULL) {
         *err = ret;
     }
-- 
2.7.4

